src/transaction/flashback.c

```diff
@@ -0,0 +1,418 @@
+
+	    /* create summary info entry, and register it to the summary list */
+	    if ((error = flashback_create_summary_entry (thread_p, &summary_entry)) != NO_ERROR)
+	      {
+		goto exit;
+	      }
+
+	    summary_entry->trid = trid;
+	    LSA_COPY (&summary_entry->end_lsa, &cur_log_rec_lsa);
+	    summary_entry->end_time = donetime->at_time;
+
+        // *INDENT-OFF*
+	context->summary_list.insert (std::make_pair (trid, summary_entry));
+        // *INDENT-ON*
+	    current_time = donetime->at_time;
+	    context->num_summary++;
+	    break;
+	  }
+	case LOG_DUMMY_HA_SERVER_STATE:
+	  {
+	    LOG_REC_HA_SERVER_STATE *ha_dummy;
+
+	    LOG_READ_ADVANCE_WHEN_DOESNT_FIT (thread_p, sizeof (*ha_dummy), &process_lsa, log_page_p);
+
+	    ha_dummy = (LOG_REC_HA_SERVER_STATE *) (log_page_p->area + process_lsa.offset);
+
+	    current_time = ha_dummy->at_time;
+
+	    break;
+	  }
+	case LOG_SUPPLEMENTAL_INFO:
+	  {
+	    LOG_REC_SUPPLEMENT *supplement;
+	    int supplement_length;
+	    SUPPLEMENT_REC_TYPE rec_type;
+	    RECDES supp_recdes = RECDES_INITIALIZER;
+
+	    OID classoid;
+
+	    LOG_READ_ADVANCE_WHEN_DOESNT_FIT (thread_p, sizeof (*supplement), &process_lsa, log_page_p);
+
+	    supplement = (LOG_REC_SUPPLEMENT *) (log_page_p->area + process_lsa.offset);
+	    supplement_length = supplement->length;
+	    rec_type = supplement->rec_type;
+
+	    LOG_READ_ADD_ALIGN (thread_p, sizeof (*supplement), &process_lsa, log_page_p);
+
+	    if (cdc_get_undo_record (thread_p, log_page_p, cur_log_rec_lsa, &supp_recdes) != S_SUCCESS)
+	      {
+		error = ER_FAILED;
+		goto exit;
+	      }
+
+	    supplement_length = sizeof (supp_recdes.type) + supp_recdes.length;
+	    if (supplement_length > supplement_alloc_length)
+	      {
+		char *tmp = (char *) db_private_realloc (thread_p, supplement_data, supplement_length);
+		if (tmp == NULL)
+		  {
+		    er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, supplement_length);
+		    error = ER_OUT_OF_VIRTUAL_MEMORY;
+		    goto exit;
+		  }
+		else
+		  {
+		    supplement_data = tmp;
+		    supplement_alloc_length = supplement_length;
+		  }
+	      }
+
+	    CDC_MAKE_SUPPLEMENT_DATA (supplement_data, supp_recdes);
+
+	    free_and_init (supp_recdes.data);
+
+	    switch (rec_type)
+	      {
+	      case LOG_SUPPLEMENT_TRAN_USER:
+		{
+		  FLASHBACK_CHECK_AND_GET_SUMMARY (context->summary_list, trid, summary_entry);
+
+		  /* user filtering */
+		  if (strlen (context->user) != 0
+		      && intl_identifier_ncasecmp (context->user, supplement_data, supplement_length) != 0)
+		    {
+		      if (summary_entry != NULL)
+			{
+			  flashback_drop_summary_entry (thread_p, summary_entry);
+                      // *INDENT-OFF*
+		      context->summary_list.erase (trid);
+                      // *INDENT-ON*
+			  context->num_summary--;
+			}
+		    }
+		  else
+		    {
+		      if (strlen (summary_entry->user) != 0)
```

Comment:
`strlen (summary_entry->user)` can be 0? It seems a fixed array so always >0

Reply:
aha, I've got it. One more question. There can be multiple LOG_SUPPLEMENT_TRAN_USER for on transaction?

Reply:
Yes, I have add comment on it. 

LOG_SUPPLEMENT_TRAN_USER is logged 
1. just before first supplemental_log is logged and 
2. just before the commit log 

So, two LOG_SUPPLEMENT_TRAN_USER logs are logged per one transaction